src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,479 @@
+  | Tpat_or (l, r, _) -> find_pattern_var l @ find_pattern_var r
+  | _ -> []
+
+let rec occuring_vars_path node =
+  let loop acc node =
+    match node.Browse_tree.t_node with
+    | Browse_raw.Expression { exp_desc = Texp_ident (path, _, _); _ } ->
+      path :: acc
+    | Pattern pat -> find_pattern_var pat @ acc
+    | _ ->
+      Lazy.force node.t_children
+      |> List.concat_map ~f:occuring_vars_path
+      |> List.append acc
+  in
+  loop [] node |> Path.Set.of_list |> Path.Set.elements |> List.rev
+  |> List.filter ~f:(fun path ->
+         (* TODO: fix this *)
+         Ident.name (Path.head path) <> "Stdlib")
+
+let analyze_expr expr expr_env ~toplevel_item =
+  let is_value_unbound path =
+    let is_bound path env =
+      try
+        let _ = Env.find_value path env in
+        true
+      with Not_found -> false
+    in
+    is_bound path expr_env && not (is_bound path toplevel_item.env)
+  in
+  let is_one_of_value_decl var_path bindings =
+    List.exists
+      ~f:(fun vb ->
+        let names = find_pattern_var vb.Typedtree.vb_pat |> Path.Set.of_list in
+        Path.Set.mem var_path names)
+      bindings
+  in
+  Browse_tree.of_node ~env:expr_env (Browse_raw.Expression expr)
+  |> occuring_vars_path
+  |> List.fold_left ~init:{ bounded_vars = []; binding_kind = Non_recursive }
+       ~f:(fun acc var_path ->
+         if is_value_unbound var_path then
+           match toplevel_item.kind with
+           | Let bindings
+             when is_recursive toplevel_item
+                  && is_one_of_value_decl var_path bindings ->
+             { acc with binding_kind = Rec_and }
+           | _ -> { acc with bounded_vars = var_path :: acc.bounded_vars }
+         else acc)
+
+let choose_name name env =
+  match name with
+  | Default { basename } -> Fresh_name.gen_val_name basename env
+  | Fixed name ->
+    if Env.bound_value name env then Fresh_name.gen_val_name name env else name
+
+let extract_to_toplevel
+    { expr;
+      expr_env;
+      name;
+      gen_binding_kind;
+      generated_binding;
+      generated_call;
+      toplevel_item;
+      call_need_parenthesis
+    } buffer =
+  let val_name = choose_name name expr_env in
+  let fresh_call =
+    let parenthised_opt s =
+      if call_need_parenthesis then "(" ^ s ^ ")" else s
+    in
+    generated_call ~name:val_name
+    |> Format.asprintf "%a" Pprintast.expression
+    |> parenthised_opt
+  in
+  let toplevel_item_span = source_sub_loc buffer toplevel_item.loc in
+  let subst_loc =
+    let start_lnum =
+      1 + expr.exp_loc.Location.loc_start.pos_lnum
+      - toplevel_item.loc.loc_start.pos_lnum
+    in
+    let end_lnum =
+      start_lnum + expr.exp_loc.loc_end.pos_lnum
+      - expr.exp_loc.loc_start.pos_lnum
+    in
+    { expr.exp_loc with
+      loc_start = { expr.exp_loc.loc_start with pos_lnum = start_lnum };
+      loc_end = { expr.exp_loc.loc_end with pos_lnum = end_lnum }
+    }
+  in
+  let substitued_toplevel_binding =
```

Comment:
```suggestion
  let substituted_toplevel_binding =
```

Reply:
Also, are these always "toplevel" ? Toplevel means an item at the root of the file. If this is the closest (structure-)item enclosing the refactoring, then it's just an (structure-)item, not a toplevel one.

Reply:
I am not sure if it is the case in the current implementation. 
@Tim-ats-d I remember that extraction was failing inside modules, is that fixed?

Reply:
Absolutely, I got a little confused.